src/compose/app.ts

```diff
@@ -631,11 +652,56 @@ export class App {
 		);
 	}
 
-	private generateContainerStep(current: Service, target: Service) {
+	// In the case where the Engine does not start the container despite the
+	// restart policy (this can happen in cases of Engine race conditions with
+	// host resources that are slower to be created but that a service relies on),
+	// we need to start the container after a delay.
+	private async requirementsMetForSpecialStart(
+		current: Service,
+		target: Service,
+	): Promise<boolean> {
+		// Shortcut the Engine inspect queries if status isn't exited
+		if (
+			current.status !== 'exited' ||
+			current.config.running !== false ||
+			target.config.running !== true
+		) {
+			return false;
+		}
+		let conditionsMetWithRestartPolicy = ['always', 'unless-stopped'].includes(
+			target.config.restart,
+		);
+
+		if (current.containerId != null && !conditionsMetWithRestartPolicy) {
+			const containerInspect = await inspectByDockerContainerId(
```

Comment:
@cywang117 We can further improve this code to use the data in the `Service` object. This probably will mean adding some additional metadata to the service class, but that is an acceptable trade-off imho. The `App` methods were previously pure functions which had a lot of benefits and adding an `async` query at this point may have undesired effects, that in addition that this can make testing harder.